### PR TITLE
Enhance schedulers to support intra-step removal of agents

### PR DESCRIPTION
**Pull request by Tortar**
As discussed in #1515 the only scheduler which checks for the presence of the agents was the BaseScheduler (and the ones which inherit the method `agent_buffer`), I suppose that this should be done also for the other schedulers except to the SimultaneousActivation one

### CODE REVIEW

**Comment by rht** [comment]
L105-107 has existed in the code (was added via #584, which didn't add the tests). I could merge this PR and add the tests later, so that people can use the feature sooner, but at this point I think we should add the tests now.

**Comment by Tortar** [comment]
Will put some tests

**Comment by Tortar** [comment]
I added some tests :-)

**Comment by Tortar** [comment]
I Is it better according to you to have a subclass of MockAgent with a different step instead of how it is done now?

File: tests/test_time.py [hunk 1]
```diff
@@ -29,7 +29,14 @@ def __init__(self, unique_id, model):
         self.steps = 0
         self.advances = 0
 
+    def kill_agent(self):
```
Comment (rht) [inline tests/test_time.py:32]: `kill_other_agent` sounds clearer. Also, I think removing just 1 agent is sufficient.

File: tests/test_time.py [hunk 2]
```diff
@@ -29,7 +29,14 @@ def __init__(self, unique_id, model):
...
+        for agent in self.model.schedule.agents:
+            if agent != self:
+                self.model.schedule.remove(agent)
+
     def stage_one(self):
+        if self.model.calling_func == "test_intrastep_remove":
```
Comment (rht) [inline tests/test_time.py:38]: Maybe define a boolean attribute `do_kill_other_agent` at agent init? `kill_other_agent` might be used in other tests.

**Comment by rht** [comment]
> @rht Is it better according to you to have a subclass of MockAgent with a different step instead of how it is done now?

I think the current implementation is fine. The step logic is still clear even with the addition of the `kill_other_agent` conditional.

File: tests/test_time.py (hunk 2 above)
Comment (rht) [reply tests/test_time.py:38]: `enable_kill_other_agent` seems better. `do_kill_other_agent` could be erroneously interpreted as a function/action (although, `enable_kill_other_agent` also does).

Comment (Tortar) [reply tests/test_time.py]: done!

**Comment by rht** [comment]
LGTM

**Pull request merged**